src/bio/smd/tests/smd_ut.c

```diff
@@ -244,9 +251,95 @@ ut_pool(void **state)
 	assert_int_equal(rc, -DER_NONEXIST);
 }
 
+static void
+ut_dev_replace(void **state)
+{
+	struct smd_dev_info	*dev_info, *tmp_dev;
+	struct smd_pool_info	*pool_info, *tmp_pool;
+	uuid_t			 dev_id3, pool_id1, pool_id2;
+	d_list_t		 pool_list, dev_list;
+	uint64_t		 blob_id;
+	int			 i, rc, pool_cnt = 0, dev_cnt = 0;
+
+	uuid_generate(dev_id3);
+	uuid_generate(pool_id1);
+	uuid_generate(pool_id2);
+
+	/* Assign pools, they were unassigned in prior pool test */
+	for (i = 0; i < 4; i++) {
+		rc = smd_pool_assign(pool_id1, i, i << 10, 100);
+		assert_int_equal(rc, 0);
+
+		rc = smd_pool_assign(pool_id2, i, i << 20, 200);
+		assert_int_equal(rc, 0);
+	}
+
+	D_INIT_LIST_HEAD(&pool_list);
+	rc = smd_pool_list(&pool_list, &pool_cnt);
+	assert_int_equal(rc, 0);
+	assert_int_equal(pool_cnt, 2);
+
+	/* Assign different blobs to dev1's targets 0, 1, 2 */
+	d_list_for_each_entry(pool_info, &pool_list, spi_link) {
+		pool_info->spi_blobs[0] = 555;
+		pool_info->spi_blobs[1] = 666;
+		pool_info->spi_blobs[2] = 777;
+	}
+
+	/* Replace dev1 with dev3 without marking dev1 as faulty */
+	rc = smd_dev_replace(dev_id1, dev_id3, &pool_list);
+	assert_int_equal(rc, -DER_INVAL);
+
+	rc = smd_dev_set_state(dev_id1, SMD_DEV_FAULTY);
+	assert_int_equal(rc, 0);
+
+	/* Replace dev1 with dev2 */
+	rc = smd_dev_replace(dev_id1, dev_id2, &pool_list);
+	assert_int_equal(rc, -DER_INVAL);
```

Comment:
maybe a comment to describe why this fails, it looks like it's because the target id list doesn't match between `dev_id1` & `dev_id2`. rc `-DER_MISMATCH` would also be more intuitive in this case.

Reply:
Right, I'd have left a comment here. The 3 tests in smd_ut.c are actually dependent, device test -> pool test -> replace test. There are two devices (dev1 and dev2) are inserted in SMD in device test, and replacing a device with another device in SMD (which means the device is used by DAOS) isn't allowed, so that's why it should return -DER_INVAL.

An faulty device can only be replaced by a new device not being used by DAOS. (not present in SMD)

Reply:
DER_MISMATCH is documented to mean a mismatch in pool group version.   We should either change that or create new errors for new issues.

Reply:
Ok, I'll change it to some other error code.

Reply:
I mean in follow-on patch.